server/app/src/main/java/com/google/heartrate/wearos/app/bluetooth/server/BluetoothServerCallback.java

```diff
@@ -0,0 +1,217 @@
+        Log.v(TAG, String.format("onNotificationSent() - status=%d", status));
+    }
+
+    /**
+     * {@link BluetoothGattServerCallback#onCharacteristicReadRequest}
+     *
+     * <p>Invoke {@link GattServiceRequestHandler#onCharacteristicRead} for given characteristic.
+     * If {@link BluetoothServer} has no handler for given characteristic or
+     * {@link GattServiceRequestHandler#onCharacteristicRead} failed,
+     * invoke {@link BluetoothServer#sendErrorResponse},
+     * otherwise {@link BluetoothServer#sendResponse} to send response to client.
+     */
+    @Override
+    public void onCharacteristicReadRequest(BluetoothDevice device, int requestId, int offset,
+                                            BluetoothGattCharacteristic characteristic) {
+        Log.v(TAG, String.format("onCharacteristicReadRequest() - device=%s characteristic=%s",
+                characteristic.getUuid(),
+                Arrays.toString(characteristic.getValue())));
+        try {
+            GattServiceRequestHandler requestHandler = mBluetoothServer.getGattServiceRequestHandler(characteristic);
+            byte[] value = requestHandler.onCharacteristicRead(device, characteristic, offset);
+
+            mBluetoothServer.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, offset, value);
+        } catch (GattException e) {
+            Log.e(TAG, String.format("onCharacteristicReadRequest() failed with exception %s", e.getMessage()));
+            mBluetoothServer.sendErrorResponse(device, requestId, e.getStatus());
+        }
+    }
+
+    /**
+     * {@link BluetoothGattServerCallback#onCharacteristicWriteRequest}
+     *
+     * <p>Invoke {@link GattServiceRequestHandler#onCharacteristicWrite} for given characteristic.
+     * If {@link BluetoothServer} has no handler for given characteristic or
+     * {@link GattServiceRequestHandler#onCharacteristicWrite} failed,
+     * invoke {@link BluetoothServer#sendErrorResponse},
+     * otherwise {@link BluetoothServer#sendResponse} to send response to client.
+     */
+    @Override
+    public void onCharacteristicWriteRequest(BluetoothDevice device, int requestId,
+                                             BluetoothGattCharacteristic characteristic, boolean preparedWrite, boolean responseNeeded,
+                                             int offset, byte[] value) {
+        Log.v(TAG, String.format("onCharacteristicWriteRequest() - device=%s characteristic=%s value=%s",
+                characteristic.getUuid(),
+                Arrays.toString(characteristic.getValue()),
+                Arrays.toString(value)));
+        try {
+            GattServiceRequestHandler requestHandler = mBluetoothServer.getGattServiceRequestHandler(characteristic);
+            requestHandler.onCharacteristicWrite(device, characteristic, offset, value);
+
+            if (responseNeeded) {
+                mBluetoothServer.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, offset, value);
```

Comment:
It might be confusing for the client to receive the written value when receiving the response. You could send `null`.